### PR TITLE
Make adding Buffers more like adding Parameters

### LINKED ISSUE #35735: Make adding buffers more like adding parameters to modules.
## 🚀 Feature
Add a `nn.Buffer` type to mirror the behavior of `nn.Parameter` without the need to explicity call `nn.Module.register_buffer`.

## Motivation

It's currently intuitive and easy to add a parameter to an `nn.Module` by wrapping it in a `nn.Parameter`. To the best of my knowledge a buffer is very similar to a parameter from an end user perspective except it doesn't get returned by `nn.Module.parameters()`.

It would therefore make sense to have a similar method for adding buffers to modules. Currently you have to explicitly call `nn.Module.register_buffer` which, in my opinion, is not very elegant as it requires you to provide the name of the member variable you want to create as a string.

    # currently
    class Foo(nn.Module):
        def __init__(self):
            super(Foo, self).__init__()
            self.weights = nn.Parameter(torch.zeros(10, 10))
            self.register_buffer('my_buffer', torch.zeros(10, 10))

    # proposed
    class Foo(nn.Module):
        def __init__(self):
            super(Foo, self).__init__()
            self.weights = nn.Parameter(torch.zeros(10, 10))
            self.my_buffer = nn.Buffer(torch.zeros(10, 10))

cc @albanD @mruberry @jbschlosser @walterddr @kshitij12345 @saketh-are

### PR DESCRIPTION
**Pull request by ekamiti**
Add similar semantics for creating a buffer object similar to creating a parameter. This is done by introducing a new Buffer class that can be used for type disambiguation. The underlying functionality of registering a buffer remains the same as the register_buffer method has not been changed. The persistent parameter in the Buffer type is to indicate whether a buffer object should be persistent or not. Other non-test changes have to do with getting the new Buffer type recognized by inductor and dynamo. Remaining changes are test changes to make sure that the Buffer type can be used as a drop in replacement for register_buffer as it just leads to register_buffer being called. The addition of this new functionality still allows for normal tensors to be used as buffers so these changes are intended to be backwards compatible.

Fixes #35735

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @tianyu-l @peterbell10

### CODE REVIEW

**Comment by ekamiti** [comment]
This had already been reviewed here https://github.com/pytorch/pytorch/pull/104069 but needed to create this due to a bad rebase that pulled too many folks in that pull request.  @lezcano

**Comment by ekamiti** [comment]
@mikaylagawarecki I opened up a new pull request due to a bad merge. Could you approve the running of all workflows for this pull request?

**Comment by Yura52** [comment]
Hi @ekamiti! Just curious, the pull request is waiting for reviews or the problem is the failed tests in CI?

**Comment by ekamiti** [comment]
> Hi @ekamiti! Just curious, the pull request is waiting for reviews or the problem is the failed tests in CI?

Not sure how I missed those thanks for bringing it up. I'll see if I can fix the failures.

**Comment by lezcano** [comment]
I am not the right person to review this PR. Perhaps @jbschlosser or @albanD...

File: torch/nn/parameter.py
Comment (albanD) [inline torch/nn/parameter.py:246]: Should this be a public atribute?

File: torch/nn/parameter.py
Comment (albanD) [inline torch/nn/parameter.py:245]: As discussed offline, we need to create a new object here, we cannot re-use data.
And .detach() is a good way to get a new object.

File: torch/nn/parameter.py
Comment (albanD) [inline torch/nn/parameter.py:263]:
```suggestion
        return f"Buffer({super().__repr__()})"
```

File: torch/nn/parameter.py
Comment (albanD) [inline torch/nn/parameter.py:265]: As discussed offline, not needed!

File: torch/_utils.py
Comment (albanD) [inline torch/_utils.py:430]: I guess this will be gone as well.

File: torch/nn/parameter.py
Comment (mikaylagawarecki) [reply torch/nn/parameter.py:263]: this one will never be used too so deleted (should we override the tensor that we return's `.__repr__`?)

File: torch/nn/parameter.py
Comment (mikaylagawarecki) [reply torch/nn/parameter.py]: Hm, what is the argument against it being public? 😅

File: test/dynamo/test_modules.py
Comment (mikaylagawarecki) [inline test/dynamo/test_modules.py:2680]: Since `nn.Buffer` calls detach, this line will no longer evaluate to True, I am not sure whether this is the intended behavior of this `mark_static` logic, or whether we should make `nn.Buffer` work with this if the tensor passed to the `nn.Buffer` constructor had previously been seen 

https://github.com/pytorch/pytorch/blob/8fe5b93667b60e37c12d288659a25cbd5ae53c79/torch/_dynamo/variables/builder.py#L1197

**Review by albanD — APPROVED** [review]
Thanks!

**Review by anijain2305 — APPROVED** [review]
Dynamo changes look good. A request would be to merge this after  https://github.com/pytorch/pytorch/pull/131275 is merged. I don't anticipate, but since buffers have special meaning for cudagraphs, I want to be extra cautious. (cc @mlazos for extra set of eyes)

But if you are in a hurry, please go ahead.

**Comment by mikaylagawarecki** [comment]
As discussed offline, will wait to merge! :)

**Review by mlazos — APPROVED** [review]

File: torch/nn/modules/module.py
Comment (dzhulgakov) [inline torch/nn/modules/module.py:1997]: This is never True because `self.register_buffer` is a bound method to self while the right hand side is an unbound method. So we always go an expensive route calling `inspect.signature`

```
>>> torch.nn.Module().register_buffer is torch.nn.Module.register_buffer
False
>>> type(torch.nn.Module()).register_buffer is torch.nn.Module.register_buffer
True
```

**Pull request closed without merging**
